File: cses/02_sorting_searching/2183_missing-coin-sum/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    coins = list(map(int, data[1:1 + n]))

    # Sắp xếp các đồng xu tăng dần để áp dụng greedy.
    coins.sort()

    # reach = tổng nhỏ nhất hiện chưa tạo được; bất biến: mọi tổng trong
    # [1, reach - 1] đều tạo được bằng các đồng đã xét.
    reach = 1
    for x in coins:
        if x <= reach:
            # Hai đoạn nối liền nhau, mở rộng phạm vi tạo được.
            reach += x
        else:
            # Đồng > reach nên có khoảng trống -> reach là đáp án nhỏ nhất bị thiếu.
            break

    sys.stdout.write(str(reach) + "\n")


if __name__ == "__main__":
    main()

File: cses/02_sorting_searching/2183_missing-coin-sum/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<long long> coins(n);
    for (long long& coin : coins) {
        cin >> coin;
    }

    // Sắp xếp các đồng xu tăng dần để áp dụng greedy.
    sort(coins.begin(), coins.end());

    // smallestMissing = tổng nhỏ nhất hiện chưa tạo được; bất biến: mọi tổng
    // trong [1, smallestMissing - 1] đều tạo được bằng các đồng đã xét.
    long long smallestMissing = 1;
    for (const long long coin : coins) {
        // Nếu đồng > smallestMissing thì có khoảng trống -> dừng, đó là đáp án.
        if (coin > smallestMissing) {
            break;
        }
        // Ngược lại hai đoạn nối liền, mở rộng phạm vi tạo được.
        smallestMissing += coin;
    }

    cout << smallestMissing << '\n';
    return 0;
}
